Add highlighting for match_only_text type

etolbakov opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Following the guidance provided in this helpful blog post([Optimize storage and performance with the MatchOnlyText field in OpenSearch](https://opensearch.org/blog/Optimize-storage-and-performance-using-MatchOnlyText-field/)), I changed the data type of several fields from "text" to "match_only_text".

However, I noticed that the highlighting functionality, which was previously working well and proved to be quite convenient, seems to be no longer available after the field type change.
I'm aware that "match_only_text" type has it's limitations (scoring, frequencies and positions, interval/span queries) but I failed to find highlighting being a tradeoff.

_OpenSearch Version: v 2.17.0_

Could you please consider this PR!
Thank you in advance,
Eugene

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->
https://github.com/elastic/elasticsearch/issues/85493

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

msfroh commented on server/src/main/java/org/opensearch/search/fetch/subphase/highlight/HighlightPhase.java:
```diff
@@ -162,7 +163,8 @@ private Map<String, Function<HitContext, FieldHighlightContext>> contextBuilders
                     if (fieldType.typeName().equals(TextFieldMapper.CONTENT_TYPE) == false
-                        && fieldType.typeName().equals(KeywordFieldMapper.CONTENT_TYPE) == false) {
+                        && fieldType.typeName().equals(KeywordFieldMapper.CONTENT_TYPE) == false
+                        && fieldType.typeName().equals("match_only_text") == false) {
```
Can we similarly make this `MatchOnlyTextFieldMapper.CONTENT_TYPE`? Thanks!

msfroh reviewed (commented):
@etolbakov  -- could you please add a test to https://github.com/opensearch-project/OpenSearch/blob/main/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/search.query/11_match_field_match_only_text.yml ? 

We have some other highlighting-specific tests (albeit with ngram token filters) over at https://github.com/opensearch-project/OpenSearch/blob/main/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/search.query/30_ngram_highligthing_field_match_only_text.yml, which may serve as examples.

It looks like your change specifically affects highlighting when you specify the fields to highlight via a wildcard pattern, right? I think if a `match_only_text` field is mentioned specifically, it has worked all along -- please let me know if I'm wrong about that. 

Can you please also add an entry to CHANGELOG.md so we remember to include this in the release notes for 2.19?

Thanks a lot!

etolbakov commented:
Hi @msfroh!
Thank you very much for the review, feedback and quick turnaround!
I'll try addressing your comments ASAP.

Regarding the usecase - yes, you are correct. 
Let me provide more details: I'm using OpenSearch Dashboards. When my field has the `text` type and the highlighting worked fine. 
In the `Inspect` -> `Response` tab, I saw the exact response from backend and every document had the `highlight` section in it.
<details>
  <summary>Sample response is hidden here ⬇️ </summary>
  
  ```json
   "took": 61,
    "timed_out": false,
    "_shards": {
      "total": 3,
      "successful": 3,
      "skipped": 0,
      "failed": 0
    },
    "hits": {
      "total": 90016,
      "max_score": null,
      "hits": [
        {
          "_index": "my_index",
          "_id": "5a008d341645f42e2b067c813ca6eb04410f50b4f9ab513cef52b29aa5067c1f",
          "_version": 1,
          "_score": null,
          "_size": 4045,
          "_source": {
            <omitted>
          },
          "fields": {
            <omitted>
          },
          "highlight": {
            "my_free_text_field": [
              <omitted>
            ]
          },
          "sort": [
            1737709464709557000
          ]
        },
    <omitted>
  ]
  ```
</details>

 As soon as I changed it to `match_only_text` the `highlight` part does not present in the response.

msfroh commented:
> I'm happy to fix that in the scope of the current PR if you don't mind.

Yes please!

etolbakov commented on modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/search.query/11_match_field_match_only_text.yml:
```diff
@@ -68,3 +68,74 @@
+"wildcard highlighting on match_only_text":
+  - skip:
+      version: " - 2.18.99"
+      reason: "wildcard highlighting on match_only_text type was added in 2.19"
```
sure, will do it!

msfroh commented on server/src/main/java/org/opensearch/search/fetch/subphase/highlight/HighlightPhase.java:
```diff
@@ -37,6 +37,7 @@
 import org.opensearch.common.regex.Regex;
 import org.opensearch.index.mapper.KeywordFieldMapper;
 import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.mapper.MatchOnlyTextFieldMapper
```
Missing semicolon. (Showing up in the build output at https://build.ci.opensearch.org/job/gradle-check/52567/console.)

If you run `./gradlew precommit`, it should catch that and any code formatting issues.

Thanks!

msfroh replied:
Nevermind -- looks like you already caught it 😁

etolbakov replied:
what a shame... really sorry about that

etolbakov replied:
Too much Python recently has dulled my Java 🙃

etolbakov commented on modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/search.query/11_match_field_match_only_text.yml:
```diff
@@ -68,3 +68,74 @@
+        body:  { "text": "the fox runs across the street" }
+        refresh: true
+
+  - do:
+      search:
+        rest_total_hits_as_int: true
+        body:
+          query:
+            match:
+              text:
+                query: fox runs
+                operator: AND
+  - match: {hits.total: 1}
+  - match: {hits.hits.0.highlight.text.0: "the <em>fox</em> <em>runs</em> across the street"}
```
Ahhh… good spot!
Thanks a lot Michael! Let me address it

etolbakov commented:
Looks like I've missed the sign-off for one of the commits, will address that later today.

msfroh commented:
Buried in the console output, I found the following:

```
FAILURE: Build failed with an exception.

* What went wrong:
Gradle build daemon has been stopped: since the JVM garbage collector is thrashing
```

I don't think that's a result of your change, so I'm going to retry the build.

etolbakov commented:
@msfroh  
Michael, thank you for your review and assistance with this PR!
Please let me know if there's anything required for backporting it

msfroh commented:
Thank you for fixing this, @etolbakov. 

It looks like the backport had a merge conflict, almost certainly on the CHANGELOG.md file, probably because we *just* cut the 2.19 branch and removed all of the pending "unreleased 2.x" lines. If you can follow the above manual backport steps ☝️ and cut a backport PR, that would be great. 

Otherwise, I'm happy to take care of it. You've done the useful work, so fighting with our branching nonsense is entirely optional. 😀

etolbakov commented:
Sure, I'll do it.
I will use PR "17210" / "17023" pair (deliberately copied ids instead of links to avoid mentioning) as an example and tag you once ready.

msfroh commented:
D'oh... it looks like it also doesn't auto-backport to the 2.19 branch. I'll take care of that one.
